simplify JALR/Sentry text for RV64 (#994)
riscv/riscv-cheri: src/cheri/insns/jalr_32bit.adoc

Reviewer: The third NOTE: is the same as the first one.

Author: ```suggestion
```

include::base_isa_extension.adoc[]

Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+
NOTE: The  description below includes behavior depending on the implemented <<sec_cap_type>> values.
 See the notes below the description.
+
. `{cs1}` is written to the target <<pcc>>.
. If `{cd}` is the link register, `{cs1}` is _sealed_ and `{cs1}` is not a _forward-edge_ <<sentry_cap>>, then set the {ctag} of the target <<pcc>> to zero.
. If `{cd}=x0`, `{cs1}` is the link register, `{cs1}` is _sealed_ and `{cs1}` is not a _backward-edge_ <<sentry_cap>>, then set the {ctag} of the target <<pcc>> to zero.
. Unseal the target <<pcc>> if it is any <<sentry_cap>>.

> . The target address is obtained by adding the sign-extended 12-bit I-immediate to `{cs1}.address`, then setting the least-significant bit of the result to zero.

Reviewer: ```suggestion
. The target address is obtained by adding the sign-extended 12-bit I-immediate to `{cs1}.address`, then setting the least-significant bit of the result to zero.
. Unseal the target <<pcc>> if it is any <<sentry_cap>> and the target address is equal to `{cs1}.address`.
```

. Set the address of the target <<pcc>> to the target address using the semantics of the <<SCADDR>> instruction.
. The <<pcc>> of the next instruction is sealed as a _backward-edge_ <<sentry_cap>> and written to `{cd}`.
. Jump to the target <<pcc>>.
+
NOTE: If the only supported <<sec_cap_type>> is zero, then no <<sentry_cap, sentry capabilities>> exist, and so _sealing_ and _unsealing_ have no effect.
 This behavior requires non-zero <<sec_cap_type>> values to be implemented, such as by the <<section_rvy_sentry_insn_ext>> extension.
+

> NOTE: When any sealed capability is passed as the input to <<JALR_CHERI>>, its address must have bit zero clear and the instruction must have a zero offset, or the target <<pcc>> will have its {ctag} set to zero, since updates to its address are interpreted with the semantics of the <<SCADDR>> instruction.

Reviewer 2: How is the tag cleared automatically when the offset is non zero and cs1 is sealed? I don't think it's specified anywhere.

Reviewer: That's what yaddrw (csetaddr) does for sealed caps

Reviewer 2: Hmm, but you work with a "target address", which is obtained after adding the offset. Shouldn't you explicitly say that if cs1 is sealed and there's an offset, then throw exception?

Author: no exception any more - we clear the tag of the target PCC - and then take an exception at the target - this was requested by ARC - it's possible to take an exception _on_ the JAL but that will need to be a new extension

Reviewer 2: Sure, exception/clear tag. But that behavior is not explained I believe, for jalr with sentry

include::base_isa_extension.adoc[]

Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+
ifndef::cheri_ratification_v1_only[]
NOTE: The  description below contains three hooks for extending <<JALR_CHERI>> behavior, used by extensions to give a large degree of extensibility.
 Unless an instruction, such as <<SENTRY>>, is implemented which explicitly reference any of the hooks, then take no action for any of them.
+
endif::[]

> . The target address is obtained by adding the sign-extended 12-bit I-immediate to the register `{cs1}`, then setting the least-significant bit of the result to zero.

Reviewer: ```suggestion
. The target address is obtained by adding the sign-extended 12-bit immediate to the register `{cs1}`, then setting the least-significant bit of the result to zero.
```

Author: the I-immediate part is taken from the normal JALR description

. The target <<pcc>> is obtained by applying the target address to `{cs1}` using the semantics of the <<SCADDR>> instruction.
ifndef::cheri_ratification_v1_only[]
. HOOK 1: Optionally clear the {ctag} of the target <<pcc>> depending on the <<sec_cap_type>> value and the numerical values of `{cd}` and `{cs1}`

> . Unseal the target <<pcc>> if it is a <<sentry_cap>>.

Reviewer: These rules are no longer quite correct. Scaddr would have detagged even when setting the address to the same value. I think maybe the easiest approach is to move unseal before scaddr but make it conditional on the address matching.

Author: I only checked the rendered version for RV64 - will check the full version

endif::[]
ifndef::cheri_ratification_v1_only[]
. Clear the {ctag} of the target <<pcc>> if it is any other sealed type.
endif::[]

> . The <<pcc>> of the next instruction is written to `{cd}`.

Reviewer: ```suggestion
. The <<pcc>> of the next instruction is sealed and written to `{cd}`.
```
Maybe a vague sealed is good enough here to allow for it being sealed as a return sentry? Since the initial spec doesn't have then we can just say is sealed and then clarify once return sentries are defined?

Author: this is the fundamental problem - without Zysentry it _can't_ be sealed - and I don't want this page to be vague in any way. I have an idea of how to fix this.....

. HOOK 1: Optionally clear the {ctag} of the target <<pcc>> depending on the <<sec_cap_type>> value and the numerical values of `{cd}` and `{cs1}`
. Unseal the target <<pcc>> if it is a <<sentry_cap>>, or clear the {ctag} if it is any other sealed type.
. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:
.. add the width of this instruction to the current <<pcc>> using the semantics of the <<SCADDR>> instructions, and
. Unseal the target <<pcc>> if it is a <<sentry_cap>>.
endif::[]
ifndef::cheri_ratification_v1_only[]
. Clear the {ctag} of the target <<pcc>> if it is any other sealed type.
endif::[]
. The <<pcc>> of the next instruction is written to `{cd}`.
ifndef::cheri_ratification_v1_only[]
.. HOOK 2: Optionally seal the return capability as a <<sentry_cap>> with a <<sec_cap_type>> defined by the implemented extensions.
. HOOK 3: Optionally make other architectural state updates.
endif::[]
. Jump to the target <<pcc>>.
+
[NOTE]
=====

> When a sealed capability is passed as the input to <<JALR_CHERI>>, its address must have bit zero clear and the instruction must have a zero offset, or the target <<pcc>> will have its {ctag} set to zero, since updates to its address are interpreted with <<SCADDR>> semantics.

Reviewer: I think this note is actually useful since otherwise you have to look at more places to understand what happens when a sentry is passed.

Author: re-instated

include::base_isa_extension.adoc[]

Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+
NOTE: The description below is generic, and implementations which need control-flow-integrity will include at least one extension which defines one or more <<sentry_cap, sentry capabilities>>.
 All such extensions are required to detail the specialized behavior of JALR to remove any ambiguity.
 For example, see <<section_rvy_sentry_insn_ext>>.
+
. `{cs1}` is written to the target <<pcc>>.

> . If `{cd}` is the link register, `{cs1}` is _sealed_ and `{cs1}` is not a _forward-edge_ <<sentry_cap>>, then the behavior is _reserved_^1^.

Reviewer: ```suggestion
. If `{cd}`≠x0, `{cs1}` is _sealed_ and `{cs1}` is not a _forward-edge_ <<sentry_cap>>, then the behavior is _reserved_^1^.
```

I think the reserved behaviour should apply to all "non-return" uses of jalr?

Author: I think that's reasonable - I guess @vmurali can confirm.

Reviewer 2: I think this is fine

. If `{cd}=x0`, '{cs1}' is the link register , `{cs1}` is _sealed_ and `{cs1}` is not a _backward-edge_ <<sentry_cap>>, then the behavior is _reserved_^1^.
. The target address is obtained by adding the sign-extended 12-bit I-immediate to `{cs1}.address`, then setting the least-significant bit of the result to zero.

> . Unseal the target <<pcc>> if it is any <<sentry_cap>> and `{cs1}.address[0]`` is zero and the I-immediate is zero.

Reviewer: ```suggestion
. Unseal the target <<pcc>> if it is any <<sentry_cap>>, `{cs1}.address[0]` is zero, and the I-immediate is zero.
```

. Set the address of the target <<pcc>> to the target address using the semantics of the <<SCADDR>> instruction.
. The <<pcc>> of the next instruction is sealed as a _backward-edge_ <<sentry_cap>> and written to `{cd}`.
. Jump to the target <<pcc>>.
+

> ^1^ Unless redefined by an extension then the _reserved_ behavior is to set the {ctag} of the target <<pcc>> to zero.

Reviewer: ```suggestion
^1^ Extensions that introduce new sealed types must define how they interact with <<JALR_CHERI>>, otherwise the behaviour is to set the {ctag} of the target <<pcc>> to zero.
```

+
[NOTE]
====

> If the only supported <<sec_cap_type>> is zero, then no <<sentry_cap, sentry capabilities>> exist, and so _sealing_ and _unsealing_ have no effect.

Reviewer 3: As above, one could support sealing without sentries.

Author: true - but not in the context of JALR? We're not trying to define JALR such that non-sentry types can be sealed.

Reviewer 3: I mean: a system could not satisfy "If the only supported <<sec_cap_type>> is zero" but still satisfy "no <<sentry_cap, sentry capabilities>> exist".  It is the latter, not the former, that means JALR doesn't seal.

> If the only supported <<sec_cap_type>> is zero, then no <<sentry_cap, sentry capabilities>> exist, and so _sealing_ and _unsealing_ have no effect.

Author: ```suggestion
If the only supported <<sec_cap_type>> is zero, then no <<sentry_cap, sentry capabilities>> exist, and so _sealing_ and _unsealing_ have no effect on JALR.
```

. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:
.. add the width of this instruction to the current <<pcc>> using the semantics of the <<SCADDR>> instruction, and
.. HOOK 2: Optionally seal the return capability as a <<sentry_cap>> with a <<sec_cap_type>> defined by the implemented extensions.
. HOOK 3: Optionally make other architectural state updates.
. Jump to the target <<pcc>>.
endif::[]
ifdef::cheri_ratification_v1_only[]
Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+
Operationally, copy `{cs1}` to the target <<pcc>> and then...
+
. Compute the target <<pcc>> address in two steps, each using the semantics of the <<SCADDR>> instruction:
.. Increment the address of the target <<pcc>> by the sign-extended 12-bit `offset`, and
.. set bit zero of the target <<pcc>> address to zero.
. Unseal the target <<pcc>> if it is a <<sentry_cap>>.
. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:
.. add the width of this instruction to the current <<pcc>> using the semantics of the <<SCADDR>> instruction, and
.. Seal the return capability as a <<sentry_cap>>.
. Jump to the target <<pcc>>.

> endif::[]

Reviewer: ```suggestion
Description::
Indirect jump to the target capability in `{cs1}` with an address offset.
+
ifndef::cheri_ratification_v1_only[]
NOTE: The  description below contains three hooks for extending <<JALR_CHERI>> behavior, used by extensions to give a large degree of extensibility.
 Unless an instruction, such as <<SENTRY>>, is implemented which explicitly reference any of the hooks, then take no action for any of them.
endif::[]
+
Operationally, copy `{cs1}` to the target <<pcc>> and then...
+
. Compute the target <<pcc>> address in two steps, each using the semantics of the <<SCADDR>> instruction:
.. Increment the address of the target <<pcc>> by the sign-extended 12-bit `offset`, and
.. set bit zero of the target <<pcc>> address to zero.
ifndef::cheri_ratification_v1_only[]
. HOOK 1: Optionally clear the {ctag} of the target <<pcc>> depending on the <<sec_cap_type>> value and the numerical values of `{cd}` and `{cs1}`
endif::[]
. Unseal the target <<pcc>> if it is a <<sentry_cap>>.
ifndef::cheri_ratification_v1_only[]
. Clear the {ctag} of the target <<pcc>> if it is any other sealed type.
endif::[]
. If `{cd}≠{creg}0`, compute the return capability and install it to `{cd}`:
.. add the width of this instruction to the current <<pcc>> using the semantics of the <<SCADDR>> instruction, and
ifndef::cheri_ratification_v1_only[]
.. HOOK 2: Optionally seal the return capability as a <<sentry_cap>> with a <<sec_cap_type>> defined by the implemented extensions.
. HOOK 3: Optionally make other architectural state updates.
endif::[]
. Jump to the target <<pcc>>.
+

NOTE: Future {cheri_base_ext_name} extensions are permitted introduce additional restrictions on the behaviour of JALR. One such example is CHERIoT differentiating between a forward and backward edge <<sentry_cap>>.
```

But I think we probably do need to keep these hooks listed since I'm not sure we can redefine behaviour later. The CHERIoT encoding can apparently redefine it but I worry about this preventing us from adding the foward-backward sentries in  later `Zysimplecfi` extension.

Maybe this suggested note is sufficient?
